### PR TITLE
DEBUG-2334 Dynamic Instrumentation Serializer component

### PR DESCRIPTION
**Pull request by p-datadog**
**What does this PR do?**

Adds the Serializer component which is responsible for converting arbitrary Ruby objects to object trees made up of basic types.

**Motivation:**

Ruby DI implementation

**Additional Notes:**

**How to test the change?**
Unit tests only in this PR.

Unsure? Have a question? Request a review!

### CODE REVIEW

File: lib/datadog/di/serializer.rb [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+            serialized.update(entries: entries)
+          end
+        else
+          if depth < 0
+            serialized.update(notCapturedReason: "depth")
+          else
+            fields = {}
+            max = settings.dynamic_instrumentation.max_capture_attribute_count
+            cur = 0
+            value.instance_variables.each do |ivar|
```
Comment (y9v) [inline lib/datadog/di/serializer.rb:142]: just out of curiosity: we are not serializing class variables since they are not frequently used?

**Review by y9v — APPROVED** [review]

File: lib/datadog/di/serializer.rb [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+      # arguments to keyword arguments ("arg1", "arg2", ...) and ensure
+      # the positional arguments are listed first.
+      def serialize_args(args, kwargs)
+        counter = 0
+        combined = args.each_with_object({}) do |value, c|
+          counter += 1
+          # Conversion to symbol is needed here to put args ahead of
+          # kwargs when they are merged below.
+          c[:"arg#{counter}"] = value
+        end.update(kwargs)
```
[the remark this answers is not in the document]
Comment (p-datadog) [reply lib/datadog/di/serializer.rb:53]: Personal preference, I suppose. I always thought `merge!` was a Rails extension for quite a while but it appears to have been added to core language in 1.8.0. `update` also does not require the use of pinky to hit the shift key.

File: lib/datadog/di/serializer.rb [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+        serialize_vars(combined)
+      end
+
+      # Serializes variables captured by a line probe.
+      #
+      # These are normally local variables that exist on a particular line
+      # of executed code.
+      def serialize_vars(vars)
+        vars.map do |k, v|
```
[the remark this answers is not in the document]
Comment (p-datadog) [reply lib/datadog/di/serializer.rb]: Changed, is this what you had in mind?

File: lib/datadog/di/serializer.rb (hunk 1 above)
Comment (y9v) [reply lib/datadog/di/serializer.rb:142]: no, I mean class variables that start with `@@` (I don't see them being used often though, due to thread-safety issues).

Rails uses them in some places:
https://github.com/rails/rails/blob/95e390156142f94649f11f803eb05cd2bb558e7d/actionview/lib/action_view/digestor.rb#L7
https://github.com/rails/rails/blob/95e390156142f94649f11f803eb05cd2bb558e7d/railties/lib/rails/railtie/configuration.rb#L9
https://github.com/rails/rails/blob/95e390156142f94649f11f803eb05cd2bb558e7d/railties/lib/rails/railtie.rb#L207

Comment (p-datadog) [reply lib/datadog/di/serializer.rb:142]: Right. What I was saying was I think it makes sense to serialize them if and only if the value being serialized is a `Class`.

Comment (y9v) [reply lib/datadog/di/serializer.rb:142]: yes, you are right. Thanks! This is all quite amazing stuff really

**Pull request merged**